LoopKit/DoseProgressTimerEstimator.swift

```diff
@@ -0,0 +1,96 @@
+//
+//  DoseProgressTimerEstimator.swift
+//  LoopKit
+//
+//  Created by Pete Schwamb on 3/23/19.
+//  Copyright © 2019 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+
+
+open class DoseProgressTimerEstimator: DoseProgressReporter {
+
+    private let lock = UnfairLock()
+
+    private var observers = WeakSet<DoseProgressObserver>()
+
+    var timer: DispatchSourceTimer?
+
+    let reportingQueue: DispatchQueue
+
+    public init(reportingQueue: DispatchQueue) {
+        self.reportingQueue = reportingQueue
+    }
+
+    open var progress: DoseProgress {
+        fatalError("progress must be implemented in subclasse")
+    }
+
+    public func addObserver(_ observer: DoseProgressObserver) {
+        lock.locked {
+            let firstObserver = observers.isEmpty
+            observers.insert(observer)
+            if firstObserver {
+                start()
+            }
+        }
+    }
+
+    public func removeObserver(_ observer: DoseProgressObserver) {
+        lock.locked {
+            observers.remove(observer)
+            if observers.isEmpty {
+                stop()
+            }
+        }
+    }
+
+    public func notify() {
+        let observersCopy = lock.locked { () -> WeakSet<DoseProgressObserver> in
```

Comment:
You should be able to remove the type annotation here, i.e. `lock.locked { observers }`